### DISCUSSION TITLE
how do I downgrade oak?

### DISCUSSION THREAD

**Discussion by ralyodio** [oakserver/oak#476] [Q&A] [answered]
The docs say to use this:

`import { Application, Router } from "https://deno.land/x/oak/mod.ts";` but it seems sometime in the past 6 months a ton of stuff broke when I upgrade deno...i don't know if its deno or oak that changed or both, I'd like to try previous version of oak first so I can get the api working again while I update my code to fix all the new errors.

I don't see a list of versions anywhere.

Here are all the new erorrs if anyone cares:

```
$ deno run --allow-all --unstable index.ts
Check file:///home/ettinger/www/grazily/grazily-api/index.ts
error: TS2707 [ERROR]: Generic type 'RouterContext<R, P, S>' requires between 1 and 3 type arguments.
  { request, response, state }: RouterContext,
                                ~~~~~~~~~~~~~
    at file:///home/ettinger/www/grazily/grazily-api/src/middleware/jwt.ts:19:33

TS2532 [ERROR]: Object is possibly 'undefined'.
    data.positive = positive.map((w) => w.word);
    ~~~~
    at file:///home/ettinger/www/grazily/grazily-api/src/controllers/searches.ts:80:5

TS2532 [ERROR]: Object is possibly 'undefined'.
    data.negative = negative.map((w) => w.word);
    ~~~~
    at file:///home/ettinger/www/grazily/grazily-api/src/controllers/searches.ts:81:5

TS2344 [ERROR]: Type 'string' does not satisfy the constraint 'unknown[]'.
      const query = db.prepareQuery<string>(
                                    ~~~~~~
    at file:///home/ettinger/www/grazily/grazily-api/src/controllers/users.ts:38:37

TS2339 [ERROR]: Property 'find' does not exist on type 'typeof Users'.
    const user = await Users.find(id);
                             ~~~~
    at file:///home/ettinger/www/grazily/grazily-api/src/controllers/users.ts:77:30

TS2532 [ERROR]: Object is possibly 'undefined'.
      users: totalUsers.pop().total,
             ~~~~~~~~~~~~~~~~
    at file:///home/ettinger/www/grazily/grazily-api/src/controllers/stats.ts:14:14

TS2532 [ERROR]: Object is possibly 'undefined'.
      searches: totalSearches.pop().total,
                ~~~~~~~~~~~~~~~~~~~
    at file:///home/ettinger/www/grazily/grazily-api/src/controllers/stats.ts:15:17

Found 7 errors.
➜  grazily-api git:(
```

**Comment by kitsonk** [comment, marked as answer]
From the readme:

> ⚠️ _Warning_ The examples in this README pull from `main`, which may not make
> sense to do when you are looking to actually deploy a workload. You would want
> to "pin" to a particular version which is compatible with the version of Deno
> you are using and has a fixed set of APIs you would expect.
> `https://deno.land/x/` supports using git tags in the URL to direct you at a
> particular version. So to use version 3.0.0 of oak, you would want to import
> `https://deno.land/x/oak@v3.0.0/mod.ts`

When using unversioned URLs with Deno is they get down into your cache and never change unless you do `--reload` or manually delete your cache.

The versions are available in the drop down on https://deno.land/x/oak/ and the changes per release are tracked in [CHANGELOG.md](https://github.com/oakserver/oak/blob/main/CHANGELOG.md).

**Comment by ralyodio** [comment]
Where can I find a list of oak versions?

**Reply by kitsonk** [reply]
Please read my replies.

> The versions are available in the drop down on https://deno.land/x/oak/ and the changes per release are tracked in [CHANGELOG.md](https://github.com/oakserver/oak/blob/main/CHANGELOG.md).

**Comment by ralyodio** [comment]
can I use /oak@9/ if i want 9.x or do i need explicit version?

**Reply by kitsonk** [reply]
`deno.land/x` currently only supports explicit versions.
